torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +24,43 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 2 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCapturePreHook
+    : GraphTask::ExecInfo::Capture::GradCapturePreHook {
+  DistAccumulateGradCapturePreHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    // It is possible that the grad is not defined since a separate
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient.
+    if (!grad.defined()) {
+      return grad;
+    }
+    autogradContext_->accumulateGrad(
+        accumulateGrad_->variable, grad, 1 /* num_expected_refs */);
+
+    if (!accumulateGrad_->post_hooks().empty()) {
+      const variable_list kEmptyOuput;
+      const variable_list inputGrads = {grad};
+      for (const auto& hook : accumulateGrad_->post_hooks()) {
+        // Discard the return value.
```

Comment:
There should be no return value here as the output is empty right?

Reply:
The post hook signature has a return value, `variable_list`. It's possible that a subclass of the base pre hook returns an non-empty `variable_list`.

https://github.com/pytorch/pytorch/blob/8544591f5a312980c353e9ff2db8a357943a3db5/torch/csrc/autograd/function_hook.h#L19-L24

Reply:
Since the hook cannot change the number or size of the gradient, you should raise an error if the hook returns anything else than an empty variable_list right?